product_state_history/models/product_template.py

```diff
@@ -0,0 +1,43 @@
+# Copyright 2020 ACSONE SA/NV (<http://acsone.eu>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import models
+
+
+class ProductTemplate(models.Model):
+    _inherit = "product.template"
+
+    def _prepare_product_state_history_values(self, state_id):
+        """Prepare product state history values for create"""
+        self.ensure_one()
+        vals = {
+            "product_template_id": self.id,
+            "product_state_id": state_id,
+        }
+        return vals
+
+    def _set_product_state_history(self, vals):
+        """Create a list of history values"""
+        history_obj = self.env["product.state.history"]
+        history_vals = []
+        state_id = vals["product_state_id"]
+        # We check if new state is different from the current one
+        for template in self.filtered(lambda t: t.product_state_id.id != state_id):
```

Comment:
it's better to add the if inside the for loop, otherwise it will loop twice

```suggestion
        for template in self:
            if template.product_state_id.id != state_id:
```